frontend/dockerfile/instructions/parse_test.go

```diff
@@ -239,3 +255,81 @@ func TestRunCmdFlagsUsed(t *testing.T) {
 	require.IsType(t, c, &RunCommand{})
 	require.Equal(t, []string{"mount"}, c.(*RunCommand).FlagsUsed)
 }
+
+func TestCopyHeredoc(t *testing.T) {
+	cases := []struct {
+		dockerfile    string
+		content       []string
+		preventExpand bool
+	}{
+		{
+			dockerfile: "COPY /foo /bar",
+			content:    nil,
+		},
+		{
+			dockerfile: "COPY <<EOF /bar\nEOF",
+			content:    []string{},
+		},
+		{
+			dockerfile: "COPY <<EOF /bar\nTESTING\nEOF",
+			content:    []string{"TESTING"},
+		},
+		{
+			dockerfile:    "COPY <<'EOF' /bar\nTESTING\nEOF",
+			content:       []string{"TESTING"},
+			preventExpand: true,
+		},
+	}
+
+	for _, c := range cases {
+		r := strings.NewReader(c.dockerfile)
+		ast, err := parser.Parse(r)
+		require.NoError(t, err)
+
+		n := ast.AST.Children[0]
+		comm, err := ParseInstruction(n)
+		require.NoError(t, err)
+		require.Equal(t, c.content, comm.(*CopyCommand).Content)
+		require.Equal(t, c.preventExpand, comm.(*CopyCommand).PreventExpand)
+	}
+}
+
+func TestRunHeredoc(t *testing.T) {
+	cases := []struct {
+		dockerfile string
+		commands   []strslice.StrSlice
+		shell      bool
+	}{
+		{
+			dockerfile: "RUN ls /",
+			commands:   []strslice.StrSlice{{"ls /"}},
+			shell:      true,
+		},
+		{
+			dockerfile: `RUN ["ls", "/"]`,
+			commands:   []strslice.StrSlice{{"ls", "/"}},
+			shell:      false,
+		},
+		{
+			dockerfile: "RUN [\"<<EOF\"]\nls /\nEOF",
+			commands:   []strslice.StrSlice{{"ls /"}},
+			shell:      true,
```

Comment:
Does this property mean it is "shell form" (not "exec" / JSON?) if so, is the `commands` value correct?

Reply:
See above, I think keeping JSON pure is probably the best way forward, so this test case in it's current form will get scrapped.